egui/src/any/serializable/type_map.rs

```diff
@@ -0,0 +1,203 @@
+pub struct TypeMap(HashMap<TypeId, AnyMapElement>);
+
+// ----------------------------------------------------------------------------
+
+impl TypeMap {
+    pub fn get<T: AnyMapTrait>(&mut self) -> Option<&T> {
+        self.get_mut().map(|x| &*x)
+    }
+
+    pub fn get_mut<T: AnyMapTrait>(&mut self) -> Option<&mut T> {
+        self.0.get_mut(&TypeId::of::<T>())?.get_mut()
+    }
+}
+
+impl TypeMap {
+    pub fn get_or_insert_with<T: AnyMapTrait>(&mut self, or_insert_with: impl FnOnce() -> T) -> &T {
+        &*self.get_mut_or_insert_with(or_insert_with)
+    }
+
+    pub fn get_or_default<T: AnyMapTrait + Default>(&mut self) -> &T {
+        self.get_or_insert_with(Default::default)
+    }
+
+    pub fn get_mut_or_insert_with<T: AnyMapTrait>(
+        &mut self,
+        or_insert_with: impl FnOnce() -> T,
+    ) -> &mut T {
+        use std::collections::hash_map::Entry;
+        match self.0.entry(TypeId::of::<T>()) {
+            Entry::Vacant(vacant) => vacant
+                .insert(AnyMapElement::new(or_insert_with()))
+                .get_mut()
+                .unwrap(), // this unwrap will never panic, because we insert correct type right now
+            Entry::Occupied(occupied) => occupied.into_mut().get_mut_or_set_with(or_insert_with),
+        }
+    }
+
+    pub fn get_mut_or_default<T: AnyMapTrait + Default>(&mut self) -> &mut T {
+        self.get_mut_or_insert_with(Default::default)
+    }
+}
+
+impl TypeMap {
+    pub fn insert<T: AnyMapTrait>(&mut self, element: T) {
+        self.0
+            .insert(TypeId::of::<T>(), AnyMapElement::new(element));
+    }
+
+    pub fn remove<T: AnyMapTrait>(&mut self) {
+        self.0.remove(&TypeId::of::<T>());
+    }
+
+    pub fn clear(&mut self) {
+        self.0.clear();
+    }
+}
+
+// ----------------------------------------------------------------------------
+
+#[test]
+fn discard_different_struct() {
+    #[derive(Clone, Debug, Serialize, Deserialize)]
+    struct State1 {
+        a: i32,
+    }
+
+    #[derive(Clone, Debug, Serialize, Deserialize)]
+    struct State2 {
+        a: String,
+    }
+
+    let file_string = {
+        let mut map: TypeMap = Default::default();
+        map.insert(State1 { a: 42 });
+        serde_json::to_string(&map).unwrap()
+    };
+
+    let mut map: TypeMap = serde_json::from_str(&file_string).unwrap();
+    assert!(map.get::<State2>().is_none());
+}
+
+#[test]
+fn new_field_between_runs() {
+    #[derive(Clone, Debug, Serialize, Deserialize)]
+    struct State {
+        a: i32,
+    }
+
+    #[derive(Clone, Debug, Serialize, Deserialize)]
+    struct StateNew {
+        a: i32,
+
+        #[serde(default)]
+        b: i32,
+    }
+
+    let file_string = {
+        let mut map: TypeMap = Default::default();
+        map.insert(State { a: 42 });
+        serde_json::to_string(&map).unwrap()
+    };
+
+    let mut map: TypeMap = serde_json::from_str(&file_string).unwrap();
+    assert!(map.get::<StateNew>().is_none());
```

Comment:
Oh right, because `TypeId` changes. In the future it would be nice to not rely on `TypeId`, but perhaps use a per-type UUID or similar to handle it. That would make persistence work across rustc versions too.